packages/tpetra/core/test/Map/Map_Bug2431.cpp

```diff
@@ -0,0 +1,347 @@
+        1997, 1998, 1999, 2000, 2001, 2002};
+    std::vector<GO> vecP3 = 
+      { 4, 5, 7, 8, 13, 14, 16, 17, 22, 23, 25, 26, 30, 34, 58, 79, 1002,
+        1018, 1019, 1020, 1039, 1040, 1041, 1957, 1975, 1976, 1978, 1996, 1997,
+        1999, 2917, 2918, 2919, 2920, 2921, 2922, 2924, 2927, 2930, 2933, 2935,
+        2936, 2937, 2938, 2939, 2940, 2941, 2942, 2943, 2944, 2945, 2948, 2951,
+        2954, 2956, 2957, 2958, 2959, 2960, 2961, 2962};
+
+    errorFlag += runTest<LO,GO,NO>("sparseTest", outStream, comm,
+                                   vecP0, vecP1, vecP2, vecP3);
+
+    // Make sure it works if some process has no data
+    std::vector<GO> empty = { } ;
+    errorFlag += runTest<LO,GO,NO>("sparseTestEmptyP2", outStream, comm,
+                                   vecP0, vecP1, empty, vecP3);
+  }
+
+  // Dense test that does not use hash tables in directory.
+  // Keep same number of IDs and structure of overlap, but
+  // narrow the range of global ID values so that processors more than 
+  // 0.1 * (max ID - min ID).
+  {
+    std::vector<GO> vecP0 = 
+      { 0, 1, 3, 4, 9, 10, 12, 13, 18, 19, 21, 22, 27, 31, 36, 37, 38, 39, 40,
+        41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58,
+        59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76,
+        77, 78, 79, 80, 81, 82};
+    std::vector<GO> vecP1 = 
+      { 1, 2, 4, 5, 10, 11, 13, 14, 19, 20, 22, 23, 28, 32, 37, 46, 55, 56,
+        58, 67, 76, 77, 79, 396, 397, 398, 399, 400, 401, 402, 404, 405,
+        406, 407, 409, 410, 413, 415, 416, 417, 418, 419, 420, 421,
+        422, 423, 424, 425, 426, 427, 428, 430, 431, 434, 436, 437,
+        438, 439, 440, 441, 442};
+    std::vector<GO> vecP2 = 
+      { 3, 4, 6, 7, 12, 13, 15, 16, 21, 22, 24, 25, 29, 33, 42, 54, 58, 59,
+        60, 75, 79, 80, 81, 557, 558, 559, 560, 561, 562, 563, 564,
+        567, 570, 571, 572, 573, 574, 575, 576, 577, 578, 579, 580,
+        581, 582, 583, 584, 585, 588, 591, 592, 593, 594, 595, 596,
+        597, 598, 599, 600, 601, 602};
+    std::vector<GO> vecP3 = 
+      { 4, 5, 7, 8, 13, 14, 16, 17, 22, 23, 25, 26, 30, 34, 58, 79, 402,
+        418, 419, 420, 439, 440, 441, 557, 575, 576, 578, 596, 597,
+        599, 617, 618, 619, 620, 621, 622, 624, 627, 630, 633, 635,
+        636, 637, 638, 639, 640, 641, 642, 643, 644, 645, 648, 651,
+        654, 656, 657, 658, 659, 660, 661, 662};
+
+    errorFlag += runTest<LO,GO,NO>("denseTest", outStream, comm,
+                                   vecP0, vecP1, vecP2, vecP3);
+
+    // Make sure it works if some process has no data
+    std::vector<GO> empty = { } ;
+    errorFlag += runTest<LO,GO,NO>("denseTestEmptyP2", outStream, comm,
+                                   vecP0, vecP1, empty, vecP3);
+  }
+
+  if (errorFlag != 0) std::cout << "End Result: TEST FAILED" << std::endl;
+  else                std::cout << "End Result: TEST PASSED" << std::endl;
+
+  return 0;
```

Comment:
Should return `EXIT_SUCCESS` on success and `EXIT_FAILURE` on failure.  That doesn't matter for the TriBITS unit test framework, but does matter if we ever change frameworks or if running the test by hand.